Frameworks/Foundation/NSNotification.mm

```diff
@@ -33,54 +51,98 @@ + (NSNotification*)notificationWithName:(NSString*)name object:(id)obj userInfo:
     return [[[self alloc] initWithName:name object:obj userInfo:info] autorelease];
 }
 
+- (Class)classForCoder {
+    return [NSNotification class];
+}
+
 /**
  @Status Interoperable
 */
-- (void)dealloc {
-    [_object release];
-    [_name release];
-    [_userInfo release];
-
-    [super dealloc];
+- (instancetype)init {
+    return NSInvalidAbstractInvocationReturn();
 }
 
 /**
  @Status Interoperable
- @Notes
 */
 - (instancetype)initWithName:(NSString*)aName object:(id)object userInfo:(NSDictionary*)info {
-    if ((self = [super init]) != nil) {
-        _object = [object retain];
-        _name = [aName copy];
-        _userInfo = [info copy];
-    }
-    return self;
+    return NSInvalidAbstractInvocationReturn();
+}
+
+/**
+ @Status Interoperable
+*/
+- (NSString*)name {
+    return NSInvalidAbstractInvocationReturn();
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
+*/
+- (id)object {
+    return NSInvalidAbstractInvocationReturn();
+}
+
+/**
+ @Status Interoperable
+*/
+- (NSDictionary*)userInfo {
+    return NSInvalidAbstractInvocationReturn();
+}
+
+static NSString* s_NSNotificationKeyName = @"NS.name";
+static NSString* s_NSNotificationKeyObject = @"NS.object";
+static NSString* s_NSNotificationKeyUserInfo = @"NS.userinfo";
+
+/**
+ @Status Interoperable
 */
 - (instancetype)initWithCoder:(NSCoder*)decoder {
-    UNIMPLEMENTED();
-    return StubReturn();
+    Class notificationClass = [self class];
+    [self release];
+
+    NSString* name = [decoder decodeObjectForKey:s_NSNotificationKeyName];
+    id object = [decoder decodeObjectForKey:s_NSNotificationKeyObject];
+    NSDictionary* userInfo = [decoder decodeObjectForKey:s_NSNotificationKeyUserInfo];
+    return [[notificationClass alloc] initWithName:name object:object userInfo:userInfo];
```

Comment:
Nah; self is simply a local that's used for the scope of this method.
`[self release]` is common in class clusters; we need to discard the instance of the toplevel class that could have been allocated and replace it with one from a factory.

Optimization opportunity: Since we know that `[self alloc]` returns an `_NSConcreteNotification`, we could set the members directly. **However**: this method gets called for all NSNotification subclasses that don't override it... so we can't do that.